Add CI job for api doc generation on master

NicolasMassart opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/teku/blob/master/CONTRIBUTING.md -->

## PR Description

After moving the gh-pages branch back in the master /docs dir, this is the next step of having the API doc automatically published, these CI changes introduce two jobs:

- the JSON extraction job, using a Teku binary built in the job and run with mock options to have the API running and extract the doc.
- the Generation of the doc version file and committing it to the branch.


## Fixed Issue(s)

fixes #1507 

## Not included in this PR

only latest version of the doc is built from master. Another PR will come for building releases (tagged versions)

NicolasMassart committed:
update CI with new api doc jobs

NicolasMassart committed:
fix comment increment issue

NicolasMassart committed:
fix comment increment issue

NicolasMassart committed:
fix comment increment issue

and add executor

NicolasMassart committed:
comment requirement to be able to test job

NicolasMassart committed:
test with require assemble

NicolasMassart committed:
use mocked teku to get the spec

NicolasMassart committed:
fix git add and add artifact so we have json on every build to check

NicolasMassart committed:
Automated update of JSON specs and version files

[skip ci] Circle CI build 24199, see https://circleci.com/gh/PegaSysEng/teku/24199

Branch: 1507_add_ci_for_api_doc, commit: 94e5ea611a7991178d2ce545df8d792ecf62c9f2, tag:none

NicolasMassart committed:
Merge branch 'master' into 1507_add_ci_for_api_doc

NicolasMassart committed:
Automated update of JSON specs and version files

[skip ci] Circle CI build 24210, see https://circleci.com/gh/PegaSysEng/teku/24210

Branch: 1507_add_ci_for_api_doc, commit: b83a50f08322bd19008b0cc66ecc9259211f374b, tag:none

joshuafernandes reviewed (approved):
LGTM

benjamincburns commented on .circleci/config.yml:
```diff
@@ -272,6 +276,71 @@ jobs:
+      - run:
+          name: Extract OpenAPI JSON
+          command: |
+            build/install/teku/bin/teku --network=mainnet --Xinterop-enabled=true --rest-api-docs-enabled=true &
```
it might be worth redirecting the output of this command to a file and including that file in the build artifacts. This will make it so that you can inspect them if Teku falls over during the build.

Something like `build/install/teku/bin/teku --network=mainnet --Xinterop-enabled=true --rest-api-docs-enabled=true 2>&1 > teku_output.log &`

and then add `teku_output.log` to your `store_artifacts` list.

benjamincburns replied:
Also, it might be worth looking into a gradle plugin that can do this at compile time rather than having to actually start a full instance of teku and grab it via the HTTP interface. I _think_ this plugin might do that for you: https://plugins.gradle.org/plugin/org.openapi.generator

NicolasMassart replied:
I already searched for this kind of plugin, but I did not find any that could fit our architecture (gradle + javalin annotations). The openapi generator one is not doing that, it's generating code and doc from spec, but it doesn't generate spec. The closest one would be https://github.com/openapi-tools/swagger-maven-plugin but it's not matching our tools for what I read.

NicolasMassart replied:
made the change on logging output

benjamincburns commented on .circleci/config.yml:
```diff
@@ -272,6 +276,71 @@ jobs:
+          command: |
+            build/install/teku/bin/teku --network=mainnet --Xinterop-enabled=true --rest-api-docs-enabled=true &
+            TEKU_PID=$!
+            wget -O openapi.json http://localhost:5051/swagger-docs
```
If wget fails because of a response error (404, 500, etc), this will still write `openapi.json`, but it won't contain what you want (it'll be empty, or it will contain the body of the error response). Also be aware that `store_artifacts` always runs, regardless if previous steps were successful. We likely want to do something like the following instead of the single `wget` line that's here now:

```bash
EXIT_CODE=0
wget -O openapi.json http://localhost:5051/swagger-docs
if [ $? != 0 ]; then
  rm openapi.json
  EXIT_CODE=1
fi

kill $TEKU_PID
exit $EXIT_CODE
```

This will cause the build to fail if the output isn't what we're expecting. There are probably more idiomatic ways to write this, however.

NicolasMassart replied:
added this and also added `--timeout=30` and  `--retry-connrefused` options to make sure it tries until service started (otherwise the command exits on connection refused directly) and to be sure of the timeout length (I did not see doc about the default value, so setting it at least indicates that we know there's a retry and timeout here and it's done on purpose)

benjamincburns commented on .circleci/config.yml:
```diff
@@ -272,6 +276,71 @@ jobs:
+          name: Commit and push doc changes
+          command: |
+            git add docs
+            git commit -m "Automated update of JSON specs and version files" \
```
I didn't know that the `-m` flag works this way. Cool!

benjamincburns commented on .circleci/config.yml:
```diff
@@ -272,6 +276,71 @@ jobs:
+            git add docs
+            git commit -m "Automated update of JSON specs and version files" \
+                       -m "[skip ci] Circle CI build ${CIRCLE_BUILD_NUM}, see ${CIRCLE_BUILD_URL}" \
+                       -m "Branch: ${CIRCLE_BRANCH}, commit: ${CIRCLE_SHA1}, tag:${CIRCLE_TAG-none}"
```
I'm not familiar with this syntax - does `${CIRCLE_TAG-none}` mean "use the contents of `CIRCLE_TAG`, or if it's empty, use `none`? If so, cool. If not, you'll likely want to create a new env var for this above like so

`[ -n "$CIRCLE_TAG" ] && TAG_NAME="$CIRCLE_TAG" || TAG_NAME="none"`

though this is probably better expressed as an if/else rather than a one-liner - it's more readable that way.

ajsutton replied:
I've always used `${CIRCLE_TAG:-none}` to mean use `none` unless `CIRCLE_TAG` is set.  Seems to work without the `:` too though.

NicolasMassart replied:
> ${parameter-default} and ${parameter:-default} are almost equivalent. The extra : makes a difference only when parameter has been declared, but is null.

https://www.tldp.org/LDP/abs/html/parameter-substitution.html

benjamincburns commented on docs/latest.json:
```diff
@@ -1 +1 @@
-{"openapi":"3.0.1","info":{"title":"Teku","description":"A minimal API specification for the beacon node, which enables a validator to connect and perform its obligations on the Ethereum 2.0 phase 0 beacon chain.","license":{"name":"Apache 2.0","url":"https://www.apache.org/licenses/LICENSE-2.0.html"},"version":"v0.11.0-dev-6c22d2ea"},"paths":{"/admin/log_level":{"put":{"tags":["Admin"],"summary" ...
\ No newline at end of file
+{"openapi":"3.0.1","info":{"title":"Teku","description":"A minimal API specification for the beacon node, which enables a validator to connect and perform its obligations on the Ethereum 2.0 phase 0 beacon chain.","license":{"name":"Apache 2.0","url":"https://www.apache.org/licenses/LICENSE-2.0.html"},"version":"v0.11.2-dev-b83a50f0"},"paths":{"/admin/log_level":{"put":{"tags":["Admin"],"summary" ...
```
If this file is generated, I don't think it should be committed. Instead, add the necessary steps to generate this file and include those as an earlier job in the build, similar to how you're generating the `openapi.json` file and persisting it via workspaces.

NicolasMassart replied:
this file is part of the docs folder, so here it was updated by running the branch CI (see commit) and left here on purpose to show it works.

benjamincburns reviewed (approved):
I'm marking this approved in principle. I wasn't totally certain if things needed to change in some of my comments, so I'll leave that up to you.

Also my mindset on reviews is that I like to point out things that are improvements, but if it's likely to be working I'll still approve. Working but okay is better than something that's not being used because it's being polished to perfection.

ajsutton reviewed (approved):
LGTM.

NicolasMassart committed:
change according to comment

and tested locally first

NicolasMassart committed:
Merge branch 'master' into 1507_add_ci_for_api_doc

NicolasMassart committed:
Automated update of JSON specs and version files

[skip ci] Circle CI build 24441, see https://circleci.com/gh/PegaSysEng/teku/24441

Branch: 1507_add_ci_for_api_doc, commit: 19c4bbd26b54f96f45be6619bf79e9981606787f, tag:none

NicolasMassart committed:
activating jobs for real on master only
